pkg/build/controller/build/build_controller.go

```diff
@@ -2023,36 +2053,168 @@ func (bc *BuildController) getAdditionalTrustedCAData(config *configv1.Image) (m
+	for _, source := range sources {
+		ds := disjointSets[source]
+		topoGraph := newTopoGraph()
+		for _, set := range *ds {
+			for i := 0; i+1 < len(set.Mirrors); i++ {
+				topoGraph.AddEdge(set.Mirrors[i], set.Mirrors[i+1])
+			}
+			sourceInGraph := false
+			for _, m := range set.Mirrors {
+				if m == source {
+					sourceInGraph = true
+					break
+				}
+			}
+			if !sourceInGraph {
+				// The build of mirrorSets guarantees len(set.Mirrors) > 0.
+				topoGraph.AddEdge(set.Mirrors[len(set.Mirrors)-1], source)
+			}
+			// Every node in topoGraph, including source, is implicitly added by topoGraph.AddEdge (there are no unconnected nodes that we would
+			// have to add separately from the edges).
+		}
+		sortedRepos, err := topoGraph.Sorted()
+		if err != nil {
+			return nil, err
+		}
+		if sortedRepos[len(sortedRepos)-1] == source {
+			// We don't need to explicitly include source in the list, it will be automatically tried last per the semantics of sysregistriesv2. Mirrors.
+			sortedRepos = sortedRepos[:len(sortedRepos)-1]
+		}
+		res = append(res, operatorv1alpha1.RepositoryDigestMirrors{
+			Source:  source,
+			Mirrors: sortedRepos,
+		})
+	}
+	return res, nil
+}
+
+func scopeMatchesRegistry(scope, reg string) bool {
+	if reg == scope {
+		return true
+	}
+	if len(scope) > len(reg) {
+		return strings.HasPrefix(scope, reg) && scope[len(reg)] == '/'
+	}
+	return false
+}
+
+func (bc *BuildController) createBuildRegistriesConfigData(config *configv1.Image, policies []*operatorv1alpha1.ImageContentSourcePolicy) (string, error) {
 	registriesConfig := config.Spec.RegistrySources
-	if len(registriesConfig.InsecureRegistries) == 0 {
+	if len(registriesConfig.InsecureRegistries) == 0 && len(policies) == 0 && config == nil {
 		klog.V(4).Info("using default insecure registry settings for builds")
 		return "", nil
 	}
-	configObj := tomlConfig{
-		Registries: registries{
-			// docker.io must be the only entry in the registry search list
-			// See https://github.com/openshift/builder/pull/40
-			Search: registryList{
-				Registries: []string{"docker.io"},
-			},
-			Insecure: registryList{
-				Registries: registriesConfig.InsecureRegistries,
-			},
-		},
+	//TODO do not know how to reconcile this with the new type
+	// docker.io must be the only entry in the registry search list
+	// See https://github.com/openshift/builder/pull/40
+
+	blockedRegs := []string{}
+	insecureRegs := []string{}
+	if config != nil {
+		insecureRegs = config.Spec.RegistrySources.InsecureRegistries
+		blockedRegs = config.Spec.RegistrySources.BlockedRegistries
+	}
+
+	//TODO the MCO code factors in the registry used for the install payload from the CVO,
+	//but seems perhaps unnecessary for builds ... in general, getting the mco and cvo objs
+	//was not a req articulated in the bug, but the `updateRegistriesConfig` I got pointed
+	//to needs data from those sources ... would rather avoid pulling those in if possible
```

Comment:
Is this referring to `getValidRegistries` making it impossible to block `ClusterVersionStatus.Desired.Image`?

If so, I agree that this check does not need to be performed. MCO checks for that so that it’s not possible to break the cluster by setting the cluster-wide blocked registries configuration, which would be disastrous; an ordinary user breaking a single build by configuring it with too strict registry block is local and easily remedied.